### ISSUE TITLE
Download/Queue counter seriously lags the app to the point it can become almost completely unresponsive

### ISSUE THREAD

**Issue by drazil100** [UnicornsOnLSD/finamp#70]
The more I try to download the slower the downloads page gets. It can get slow enough the back button doesn't respond and you cant actually scroll the page. If I scroll down far enough on the downloads page that I assume the counter unloads the page scrolls super smoothly again but as soon as I scroll back up it starts getting choppy again and as soon as I leave the downloads page it performs smoothly again also.

I am running iOS version 0.5.0

**Comment by drazil100** [comment]
https://youtu.be/ypZtGg2O320

**Comment by UnicornsOnLSD** [comment]
The method I'm using to update the counter at the tip is very inefficient. Every time something changes in a download, it has to run through every download and count up how many are complete/downloading etc. I've thought of a new way of updating download counters which should be much more efficient and should fix the issue with download indicators not properly updating. I haven't actually implemented it yet, but I should be able to do it for the next release.

**Comment by UnicornsOnLSD** [comment]
This should be improved in 2fb2207, which uses the new system for tracking downloads. Instead of counting through the list 5 times every time a download update is pushed (which would be 10000+ calculations for you), the list is looped through once (when the overview is first shown) and updates are checked incrementally and are basically just adding 1 to a variable. I'll leave this issue open just in case this doesn't fix it for you. Also, the app does generally get a bit janky when downloading for some reason. I think this is caused by the downloader package I'm using, but it may be my code too.

**Comment by drazil100** [comment]
Ok the queue performs WAY better... But adding songs to the queue now lags the ENTIRE app (not just the downloads page). Once all songs are enqueued or running the app performs perfectly with no issues whatsoever.

**Comment by UnicornsOnLSD** [comment]
Could you send a video? I usually get a lot of jitter when lots of downloads are being processed, might actually start looking for what caused it since it's always annoyed me

**Comment by drazil100** [comment]
1712 is the magic number. I had 1221 songs downloaded and I am trying to queue up 491 songs. Once "Downloads" reaches 1712 and all songs that are to be downloaded are either already downloaded, enqueued, or running it starts scrolling smoothly again. 

https://youtu.be/VdlKxIP5uu0

**Comment by UnicornsOnLSD** [comment]
Yeah, that's the jitter I've been seeing. Weird that it stopped after 1712 downloads though

**Comment by drazil100** [comment]
My assumption is it has to do with it fetching the data or creating objects or something. Again once the total "downloads" number catches up to the actual number of songs that will be on your device it downloads them with no lag whatsoever. It's just populating the queue where it lags.

**Comment by UnicornsOnLSD** [comment]
That makes sense, didn't realise that downloads were still being added in the video
